clang/lib/Sema/SemaTemplateInstantiateDecl.cpp

```diff
@@ -6193,7 +6193,13 @@ static void processSYCLKernel(Sema &S, FunctionDecl *FD, MangleContext &MC) {
   if (S.LangOpts.SYCLIsDevice) {
     S.ConstructOpenCLKernel(FD, MC);
   } else if (S.LangOpts.SYCLIsHost) {
-    CXXRecordDecl *CRD = (*FD->param_begin())->getType()->getAsCXXRecordDecl();
+    QualType KernelParamTy = (*FD->param_begin())->getType();
+    CXXRecordDecl *CRD;
+    if (KernelParamTy->isReferenceType())
+      CRD =
+          const_cast<CXXRecordDecl *>(KernelParamTy->getPointeeCXXRecordDecl());
+    else
+      CRD = KernelParamTy->getAsCXXRecordDecl();
```

Comment:
I don't think that `const_cast` is a good idea.
```suggestion
    const CXXRecordDecl *CRD;
    CRD = (KernelParamTy->isReferenceType())? KernelParamTy->getPointeeCXXRecordDecl()) :
      KernelParamTy->getAsCXXRecordDecl();
```

Reply:
I may have led @smanna12 down the const_cast path.   The question is if CRD is ptr-to-const, would it still be valid to extract and modify methods accessed off it?

Reply:
We will be able to access methods of it, not sure if it will be possible to modify them. Do we really need to? I didn't see why we can't use `const` qualifier because we don't modify anything, just access, and it seems my suggestion worked.